### PR TITLE
feat: Add /health endpoint for server monitoring

### PR DESCRIPTION
**Pull request by lfelipediniz**
## 🚑 Add Health Check Endpoint to Dash Core (#3446)

This PR introduces a built-in `/health` endpoint for Dash applications, designed to support load balancers, Docker health checks, Kubernetes probes, monitoring scripts, and CI/CD pipelines.

### ✨ Key Features

* Adds `/health` endpoint to all Dash apps
* Configurable path support (respects `routes_pathname_prefix`)
* Toggle to disable endpoint if needed
* Includes detailed metadata:

  * Dash version, Python version, platform
  * Server info (host, port, debug mode)
  * System metrics (CPU, memory, disk usage via `psutil`)
  * Callback counts (total and background)
*  Includes comprehensive test suite with 10 test cases
* Endpoint is unauthenticated by default to support external probes

### ✅ Example Output

```json
{
  "status": "healthy",
  "timestamp": "2025-09-30T00:35:03.982409Z",
  "dash_version": "3.2.0",
  "python_version": "3.12.3",
  "platform": "Linux-6.8.0-71-generic-x86_64",
  "server_name": "my_dash_app",
  "debug_mode": true,
  "host": "127.0.0.1",
  "port": "8050",
  "system": {
    "cpu_percent": 17.8,
    "memory_percent": 37.9,
    "disk_percent": 52.6
  },
  "callbacks": {
    "total_callbacks": 4,
    "background_callbacks": 0
  }
}
```

### 🔧 Configuration Notes

* The endpoint automatically adjusts to any `routes_pathname_prefix`
* It is designed to be fast and lightweight (no callback triggering)
* `psutil` is optional - system metrics are included only if available

### 🔬 Testing

A full suite of 10 tests has been added, covering:

* Basic response
* Disabled endpoint
* Custom prefix support
* Missing `psutil`
* System info fallback
* Edge cases with callback registration
* Status code and JSON format

All tests pass locally :)

### CODE REVIEW

**Comment by ndrezn** [comment]
cc @selfeki

File: dash/dash.py [hunk 1]
```diff
@@ -466,6 +470,7 @@ def __init__(  # pylint: disable=too-many-statements
         description: Optional[str] = None,
         on_error: Optional[Callable[[Exception], Any]] = None,
         use_async: Optional[bool] = None,
+        health_endpoint: Optional[str] = "health",
```
[the remark this answers is not in the document]
Comment (T4rk1n) [reply dash/dash.py:473]: I agree it should default to none, but also should be automatically added if running on dash-enterprise/cloud, maybe with a variable like the other configs.

File: dash/dash.py [hunk 2]
```diff
@@ -975,6 +983,70 @@ def serve_reload_hash(self):
...
+        
+        Returns a JSON response indicating the server is running and healthy.
+        This endpoint can be used by load balancers, monitoring systems, 
+        and other platforms to check if the Dash server is operational.
+        
+        :return: JSON response with status information
+        """
+        import datetime
+        import platform
+        import psutil
```
Comment (T4rk1n) [inline dash/dash.py:998]: I don't think there should be any code in the default handler, it should simply returns `"OK"` with response 200. The api should provide a way for the developer to add a function to handle the response and do it's health check properly by checking the needed services.

File: dash/dash.py [hunk 3]
```diff
@@ -975,6 +983,70 @@ def serve_reload_hash(self):
...
+            health_data.update({
+                "callbacks": {
+                    "total_callbacks": len(self.callback_map),
+                    "background_callbacks": len(getattr(self, '_background_callback_map', {})),
+                }
+            })
+        except Exception:
+            pass
+        
+        return flask.jsonify(health_data)
```
Comment (T4rk1n) [inline dash/dash.py:1048]: Normal health check should return "OK", 200

**Review by scjody — COMMENTED** [review]
This is a great idea! I'll let those more familiar with the codebase give a detailed view.

I'd like to see a `/ready` endpoint as well. Can one be added as part of this work? It should be possible to reuse a lot of this code. To summarize the difference:
* Health checks ("is this container responding _at all_?") is relevant to the cluster scheduler.
* Readiness checks ("is this container able to respond to _this class of requests_ right now?") are relevant to the load balancer.

For example, if a service has an external database connection, one might want to have the health check just make sure the app can respond to traffic, but have the readiness check test that the DB is reachable and can perform a basic query.

File: dash/dash.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ndrezn) [reply dash/dash.py:473]: As long as we can enable it by default on Cloud, then 👍 for disabled by default

File: dash/dash.py [hunk 4]
```diff
@@ -975,6 +983,70 @@ def serve_reload_hash(self):
...
+        # Add callback information
+        try:
+            health_data.update({
+                "callbacks": {
+                    "total_callbacks": len(self.callback_map),
+                    "background_callbacks": len(getattr(self, '_background_callback_map', {})),
+                }
+            })
+        except Exception:
+            pass
```
[the remark this answers is not in the document]
Comment (ndrezn) [reply dash/dash.py:1046]: Based on some other comments I think it's probably better to break this out into a separate PR to decide how it should be exposed
